src/edge_index.cpp

```diff
@@ -54,78 +55,33 @@ const HighFive::Group targetIndex(const HighFive::Group& h5Root) {
-        // implementation
-        // TODO: throw a SonataError instead
-        return Selection({});
-    }
-
-    RawIndex primaryRange;
-    indexGroup.getDataSet(NODE_ID_TO_RANGES_DSET)
-        .select({static_cast<size_t>(nodeID), 0}, {1, 2})
-        .read(primaryRange);
-
-    const uint64_t primaryRangeBegin = primaryRange[0][0];
-    const uint64_t primaryRangeEnd = primaryRange[0][1];
-
-    if (primaryRangeBegin >= primaryRangeEnd) {
-        return Selection({});
-    }
-
-    RawIndex secondaryRange;
-    indexGroup.getDataSet(RANGE_TO_EDGE_ID_DSET)
-        .select({static_cast<size_t>(primaryRangeBegin), 0},
-                {static_cast<size_t>(primaryRangeEnd - primaryRangeBegin), 2})
-        .read(secondaryRange);
-
-    Selection::Ranges ranges;
-    ranges.reserve(secondaryRange.size());
-
-    for (const auto& row : secondaryRange) {
-        ranges.emplace_back(row[0], row[1]);
-    }
-
-    return Selection(std::move(ranges));
-}
-
 Selection resolve(const HighFive::Group& indexGroup, const std::vector<NodeID>& nodeIDs) {
-    constexpr size_t min_gap_size = SONATA_PAGESIZE / (2 * sizeof(uint64_t));
-    constexpr size_t max_aggregated_block_size = 128 * min_gap_size;
-
-    if (nodeIDs.size() == 1) {
-        return resolve(indexGroup, nodeIDs[0]);
-    }
-
-    auto readBlock = [&indexGroup](auto& buffer, const auto& range, const std::string& dset_name) {
-        size_t i_begin = std::get<0>(range);
-        size_t i_end = std::get<1>(range);
-
-        indexGroup.getDataSet(dset_name).select({i_begin, 0}, {i_end - i_begin, 2}).read(buffer);
-    };
-
-    auto primaryRange = bulk_read::bulkRead<std::array<uint64_t, 2>>(
-        [&readBlock](auto& buffer, const auto& range) {
-            readBlock(buffer, range, NODE_ID_TO_RANGES_DSET);
-        },
-        Selection::fromValues(nodeIDs),
-        min_gap_size,
-        max_aggregated_block_size);
-
-    // Sort and eliminate empty ranges.
+    auto node2ranges_dset = indexGroup.getDataSet(NODE_ID_TO_RANGES_DSET);
+    auto node_dim = node2ranges_dset.getSpace().getDimensions()[0];
+    auto sortedNodeIds = nodeIDs;
+    bulk_read::detail::erase_if(sortedNodeIds, [node_dim](auto id) {
+        // Filter out `nodeIDs[i] >= dims`; because SYN2 used to return an
+        // empty range for an out-of-range `nodeId`s.
+        return id >= node_dim;
+    });
+    std::sort(sortedNodeIds.begin(), sortedNodeIds.end());
+    sortedNodeIds.erase(std::unique(sortedNodeIds.begin(), sortedNodeIds.end()),
+                        sortedNodeIds.end());
+
+    auto nodeSelection = Selection::fromValues(sortedNodeIds);
+    auto primaryRange = detail::readCanonicalSelection<std::array<uint64_t, 2>>(
+        node2ranges_dset, nodeSelection.ranges(), RawIndex{{0, 2}});
+
+    bulk_read::detail::erase_if(primaryRange, [](const auto& range) {
+        // Filter out any invalid ranges `start >= end`.
+        return range[0] >= range[1];
```

Comment:
how hard would it be to have a test for this? I see that we're kinda lacking the `tests/test_edges.cpp` front, but maybe it's a chance to start adding some?

Reply:
Isn't this already tested here:
https://github.com/BlueBrain/libsonata/blob/bfd817f3c5a939bed92e826f01993b6305bad263/tests/test_edges.cpp#L49
https://github.com/BlueBrain/libsonata/blob/bfd817f3c5a939bed92e826f01993b6305bad263/tests/test_edges.cpp#L55

The indices are:
```
$ h5dump -d edges/edges-AB/indices/target_to_source/node_id_to_ranges tests/data/edges1.h5 
HDF5 "tests/data/edges1.h5" {
DATASET "edges/edges-AB/indices/target_to_source/node_id_to_ranges" {
   DATATYPE  H5T_STD_U64LE
   DATASPACE  SIMPLE { ( 4, 2 ) / ( 4, 2 ) }
   DATA {
   (0,0): 0, 1,
   (1,0): 1, 3,
   (2,0): 3, 5,
   (3,0): 5, 5
   }
}
}

$ h5dump -d edges/edges-AB/indices/source_to_target/node_id_to_ranges tests/data/edges1.h5
HDF5 "tests/data/edges1.h5" {
DATASET "edges/edges-AB/indices/source_to_target/node_id_to_ranges" {
   DATATYPE  H5T_STD_U64LE
   DATASPACE  SIMPLE { ( 4, 2 ) / ( 4, 2 ) }
   DATA {
   (0,0): 0, 0,
   (1,0): 0, 1,
   (2,0): 1, 2,
   (3,0): 2, 3
   }
}
}
```
Meaning `0` and `3` are empty ranges.

Reply:
I suppose; since the functions can be called w/o an hdf5 file, it would probably make it easier to make changes in the future if they had explicit tests, but we can skip that for now.